### PR TITLE
JAMES-3539 Add PushSubscription set/destroy method

### PR DESCRIPTION
**Pull request by Arsnael**

### CODE REVIEW

**Review by chibenwa — APPROVED** [review]

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/PushSubscriptionSetDeletePerformer.scala [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        .toMap
+  }
+}
+
+class PushSubscriptionSetDeletePerformer @Inject()(pushSubscriptionRepository: PushSubscriptionRepository) {
+  def deletePushSubscriptions(pushSubscriptionSetRequest: PushSubscriptionSetRequest, mailboxSession: MailboxSession): SMono[PushSubscriptionDeletionResults] =
+    SFlux.fromIterable(pushSubscriptionSetRequest.destroy.getOrElse(Seq()))
+      .flatMap(unparsedId => delete(unparsedId, mailboxSession)
+        .onErrorRecover(e => PushSubscriptionDeletionFailure(unparsedId, e)),
+        maxConcurrency = 5)
```
Comment (vttranlina) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/PushSubscriptionSetDeletePerformer.scala:64]: I remember we have a common const for this

Comment (Arsnael) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/PushSubscriptionSetDeletePerformer.scala:64]: Looked in the code by doing a search on `maxConcurrency`, every occurrence is defined like this. Where is that const???

Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/PushSubscriptionSetDeletePerformer.scala:64]: We could IMO use ReactorUtils.DEFAULT_CONCURRENCY

File: server/protocols/jmap-rfc-8621-integration-tests/distributed-jmap-rfc-8621-integration-tests/src/test/java/org/apache/james/jmap/rfc8621/distributed/DistributedPushSubscriptionSetMethodTest.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                .deduplication()
+                .noCryptoConfig())
+            .build())
+        .extension(new DockerElasticSearchExtension())
+        .extension(new CassandraExtension())
+        .extension(new RabbitMQExtension())
+        .extension(new AwsS3BlobStoreExtension())
+        .server(configuration -> CassandraRabbitMQJamesServerMain.createServer(configuration)
+            .overrideWith(new TestJMAPServerModule(), new PushSubscriptionProbeModule()))
+        .build();
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/distributed-jmap-rfc-8621-integration-tests/src/test/java/org/apache/james/jmap/rfc8621/distributed/DistributedPushSubscriptionSetMethodTest.java:54]: Careful we now need the Mock HTTP server extension here...

**Comment by chibenwa** [comment]
Rebase please (And check for rebase conflicts)

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/PushSubscriptionSetDeletePerformer.scala (hunk 1 above)
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/PushSubscriptionSetDeletePerformer.scala:64]: Should we? ReactorUtils.DEFAULT_CONCURRENCY = 16... not 5.

What is the impact?

**Comment by Arsnael** [comment]
Rebased

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/PushSubscriptionSetDeletePerformer.scala (hunk 1 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/PushSubscriptionSetDeletePerformer.scala:64]: I'm OK keeping it as it is.

**Review by chibenwa — APPROVED** [review]

**Comment by Arsnael** [comment]
```
03:52:57,295 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-install-plugin:2.5.2:install (default-install) on project james-server-spring-app: Failed to install metadata org.apache.james:james-server-spring-app:3.7.0-SNAPSHOT/maven-metadata.xml: Could not parse metadata /home/jenkins/.m2/repository/org/apache/james/james-server-spring-app/3.7.0-SNAPSHOT/maven-metadata-local.xml: in epilog non whitespace content is not allowed but got c (position: END_TAG seen ...</metadata>\nc... @21:2) -> [Help 1]
```
Here starts the battle against the CI ><

**Review by vttranlina — APPROVED** [review]

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/PushSubscriptionSetDeletePerformer.scala [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import reactor.core.scheduler.Schedulers
+
+import javax.inject.Inject
+
+object PushSubscriptionSetDeletePerformer {
+  sealed trait PushSubscriptionDeletionResult
+  case class PushSubscriptionDeletionSuccess(PushSubscriptionId: PushSubscriptionId) extends PushSubscriptionDeletionResult
+  case class PushSubscriptionDeletionFailure(PushSubscriptionId: UnparsedPushSubscriptionId, exception: Throwable) extends PushSubscriptionDeletionResult {
+    def asPushSubscriptionSetError: SetError = exception match {
+      case e: PushSubscriptionNotFoundException => SetError.notFound(SetErrorDescription(e.getMessage))
```
Comment (quantranhong1999) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/PushSubscriptionSetDeletePerformer.scala:38]: Why do we need this notFound ex when revoke?

Comment (Arsnael) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/PushSubscriptionSetDeletePerformer.scala:38]: Hmmm good point actually, as we said the revoke is idempotent, we likely never get this exception being thrown

Comment (Arsnael) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/PushSubscriptionSetDeletePerformer.scala]: Fixed (but I had a green build finally ><)

**Review by quantranhong1999 — APPROVED** [review]

**Pull request merged**